client/executor/wasmtime/src/state_holder.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::host::{HostContext, HostState};
+use std::cell::RefCell;
+use std::rc::Rc;
+
+/// A common place to store a reference to the `HostState`.
+///
+/// This structure is passed into each host function handler and retained in the implementation of
+/// `WasmRuntime`. Whenever, a call into a runtime method is initiated the host state is populated
+/// with the state for that runtime method call.
+///
+/// During the execution of the runtime method call, wasm can call imported host functions. When
+/// that happens the host function handler gets a `HostContext` (obtainable through having a
+/// `HostState` reference).
+#[derive(Clone)]
+pub struct StateHolder {
+	// This is `Some` only during a call.
+	state: Rc<RefCell<Option<HostState>>>,
+}
+
+impl StateHolder {
+	/// Create a placeholder `StateHolder`.
+	pub fn empty() -> StateHolder {
+		StateHolder {
+			state: Rc::new(RefCell::new(None)),
+		}
+	}
+
+	/// Provide `HostState` for the runtime method call and execute the given function `f`.
+	///
+	/// During the execution of the provided function `with_context` will be callable.
+	pub fn init_state<R, F>(&self, state: &mut HostState, f: F) -> R
+	where
+		F: FnOnce() -> R,
+	{
+		// We need an `Option` here since Rust will doesn't allow to assign a value for the first
+		// time from within a closure.
+		let mut ret: Option<R> = None;
+
+		take_mut::take(state, |state| {
```

Comment:
Hmm, I think that will not work. It internally catches panics and aborts the process afterwards, but we can not abort the process.

Reply:
Ah, good catch!